README.md
# Cjax Web Development Ajax Framework

## Requirements

+    PHP 5.4+
+    javascript enabled
+    curl, eval, ini_set functions enabled


## Links

 Official Site: http://cjax.sourceforge.net/

 Official Repo: https://github.com/ajaxboy/cjax

 Official Download Location: https://sourceforge.net/projects/cjax/files/


## Instructions

This reposition is for Fork and contributions only, for downloads go to: http://cjax.sourceforge.net/. We build the releases internally because there are different packages that derive from this project (examples, plugins, docs, and 2 versions of Cjax).

## Introduction

Cjax is a small lightweight but very powerful Ajax Framework that translate PHP Code into User Interface Actions. It can 
help to develop Ajax applications in record time, saving you up to 60% of development time once you are familiar with the
simple to use Cjax API.
  
Cjax is simplistic and uses Convention  Over Configuration model approach. With its MVC design makes it quite easy to integrate
into existing PHP-Frameworks (such as Zend, CodeIgnater, cakePHP, Yii, etc).

In most cases all you need is one line of code. Cjax carries a plugin system (as well as the core) that makes PHP and JavaScript look like lovers, meaning that
it is the most seemless integration of PHP and JavaScript that you can find.

Unlike others libraries that try to mimic Cjax, for cjax there is no 'client' side involved, there is no inlines codes, there is no process request, all you need is your php code and that's it.

Cjax is so easy to use that even a 12 year old could figure it out, go ahead, give it a try.

## Example code

Check the a lot of examples and code at http://cjax.sourceforge.net/examples/

## Full API Documentation

http://cjax.sourceforge.net/docs/ (Also anyone can contribute/folk to the docs on this repo).

## Documentation

It's really simple as

## The CJax's way, one line of php code does it all

## Example #1 - Ajax Call
```php
<?php
$ajax->click('#element_id' , $ajax->call('controller/the_function/'));
?>
<!doctype html>
<html>
	<head>
		<?php echo $ajax->init();?>
	</head>
	<body>
		<a id='element_id' href='#'>Click me</a>
	</body>
</html>
```

## Example #2 - Ajax Form
```php
<?php
$ajax->click('#element_id' , $ajax->form('controller/the_function/'));
?>
<!doctype html>
<html>
	<head>
		<?php echo $ajax->init();?>
	</head>
	<body>
	<form>
		Field #1 <input type='text' id='field1' name='field1' />
		<br />
		Field #2 <input type='text' id='field1' name='field2' />
		<br />
		<a id='element_id' href='#'>Click me To Submit Form</a>
	</form>
	</body>
</html>
```

## Example #3 - Ajax Overlay, Lightbox
```php
<?php
$ajax->click('#element_id' , $ajax->overlay('some/splash/html'));
?>
<!doctype html>
<html>
	<head>
		<?php echo $ajax->init();?>
	</head>
	<body>
		<a id='element_id' href='#'>Click Show Overlay</a>
	</body>
</html>
```

## 50+ more examples here: http://cjax.sourceforge.net/examples/


CHANGELOG.md

# ChangeLog

## CJAX 6.0 - TBD

- Namespace, closure and short array syntax used in the framework, PHP version now requires 5.4.0+
- Fixed strict standard errors resulted from calling instance methods statically for PHP 5.5 and 5.6 compatibility
- Added autoloader for CJAX classes, no need for require/include statements
- Added base abstract AJAX Controller class, every concrete AJAX controller class should extend this class
- Added functionality to pass javascript event object to PHP through cookie
- Added event classes that represent valid javascript event objects to use in AJAX controllers
- Added support for Mouse events
- Added support for Drag-Drop events
- Added plugin Dracula(Dragula) for CJAX
- Added documentation for CJAX API in PHPDocumentator style
- Updated folder structure for CJAX
- Updated examples and Added new examples for Mouse and Drag-Drop events
- Updated readme.md, changelog.md
- Updated versions for third party libraries such as sizzle
- Removed unused and deprecated files, classes and code
- Other fixes and improvements.


## CJAX 5.8 - Oct 6, 2013

- Updated examples and readme.md.
- Other fixes and improvements.


## CJAX 5.7 - Oct 3, 2013

- Advanced selectors (Sizzle, JQuery selectors allowed)
- Fixed issue when passing multiple url values
- Fixed issue in processing that was introduced in previous release
- Added support for main properties for tables tr,td,thead,tbody, etc.


## CJAX 5.6 - September 19, 2013

- Fixed issue displaying a blank page some times (Thanks Stefan for reporting this)
- Fixed click trigger event
- Improved some Js access functions processing
- Added new Main property in samples
- Added Js Properties in samples
- Added Location in samples
- Added click() on samples
- Fixed form loading text option not display when submitting a form (thanks Jim for reporting this)
- Plugin uploader now included in main release package
- Plugin uploadify now included in main release package
- Plugin validate now included in main release package
- Plugin autocomplete now included in main release package
- Plugin ExamplePlugin now included in main release package


## CJAX 5.5 - April 10, 2013

- Support for PHP5.4x
- Fixed plugins docs issues with an old function reference interrupting the page from displaying.

 
## CJAX 5.4 - February 23, 2013

- Support for keypress event
- Ability for url element values to process Js modifiers and functions.
- Fixed minor problems with plugin data processor
- Enhanced some range statements
- Removed autoloading function
- Added extra PHP settings check to check for error display
- Added Strict Standard Check and tries to automatically turn it off if it is on.
- Added PHP version check, so it shows a message if your php versions is not compatible.
- Fixed some notices
- Enhanced the uploader and made more efficient and reliable. Re-did the ajax uploader documentation
- Other fixes and improvements
	
	
## CJAX 5.3-Obama2012(Maintenance Release) - Nov 2nd, 2012

- New Autocomplete plugin
- Ability for wait() function to process milliseconds
- Implemented new release "code name" policy: 
  - This means that a code name will be used for each release or the same code name for a series of releases.
- Uploadify plugin was updated to uploadify 3.2
- Fixed bug that cut off parameters if two slashes were passed in the URL.
- Other improvements & small fixes

## CJAX. 5.2-Stable(Maintenance Release) - October 29, 2012

- Support to change table element
- Improved Select API, now can specify a default value
- Changed AJAX_CD cookie deletion
- Added Support for Exec events on Plugins.

## CJAX 5.1-Stable(Maintenance Release) - September 16, 2012

- New usage for Exec events:

  ```php
	$ajax->click($element_id, $actions);
	$ajax->change($element_id, $actions);
	$ajax->keyup($element_id, $actions);
	$ajax->keydown($element_id, $actions);
	$ajax->blur($element_id, $actions);
  ```
	
	This is alternative replacement wrapper of the clasic:

  ```php	
	$ajax->Exec();
  ```

- Multiple APIs now work on overlays callbacks for example you may use Cjax in overlays elements.
 
  Eg:
  ```php  
	$overlay = $ajax->overlay('resources/html/login.html');
	
	//work within the scope of the overlay (elements in resources/html/login.html)
	
	//would change the title of the page to whatever you type in a text field on the overlay!.
	$overlay->keyup('field1', $ajax->document('title','|field1|'));
	
	//submit ajax-form in the overlay
	$overlay->click('button1', $ajax->form('ajax.php?ajax_login/handler'));
	
	//assign the overlay to an element so when you click on it, the overlay pops
	$ajax->click('a_login', $overlay);
  ```	
  On 5.0, this was limited to 1 single API. Now you may add as many as you want.

- Minor improvements & minor bugs fixed
- Expanded the number of parameters you can pass from a to f to a to z.
- Fixed "select" bug that chopped off last item in the array list in "select" ajax API.
- Added support for PHP 5.2
- Allow to require 'ajax.php' for codeIgnater, when before it had to be 'ajaxfw.php', (the old way still works)
- Validate & Uploader plugins integration
  Plugin validate now integrates with plugin uploader.
 
  Example:

  ```php  
   $ajax->validate('buttonID', 'ajax.php?optional/pre/action', $rules)->uploader(
	   ['target' => 'directory/where/files/are/uploaded/to',
		'text'=> 'Uploading Files..',
		'url' => 'ajax.php?optional/post/action', //form post after the files are uploaded
		'form_id' => 'form_id',
		'success_message' => 'Files Uploaded Successfully..',
		'suffix' => time()]
	);
  ```
  

## CJAX 5.0-Stable Gold - July 2, 2012

- Minor improvements and semantic improvements were made, and this marks the roll out of the official stable release.
	

## CJAX 5.0-RC3 - Jun 30, 2012

- Fixed some small bugs introduced in RC2	

cjax/core/cjax.php
<?php

/** ################################################################################################**   
* Copyright (c)  2008  CJ.   
* Permission is granted to copy, distribute and/or modify this document   
* under the terms of the GNU Free Documentation License, Version 1.2   
* or any later version published by the Free Software Foundation;   
* Provided 'as is' with no warranties, nor shall the author be responsible for any misuse of the same.     
* A copy of the license is included in the section entitled 'GNU Free Documentation License'.   
*   
*   CJAX  6.0               $     
*   ajax made easy with cjax                    
*   -- DO NOT REMOVE THIS --                    
*   -- AUTHOR COPYRIGHT MUST REMAIN INTACT -   
*   Written by: CJ Galindo                  
*   Website: http://cjax.sourceforge.net                     $      
*   Email: cjxxi@msn.com    
*   Date: 2/12/2007                           $     
*   File Last Changed:  04/06/2016           $     
**####################################################################################################    */   

namespace CJAX\Core;

class CJAX extends Framework{

	private static $CJAX;
		
        
	/**
	* get an instance of CJAX
	* with singleton pattern 
	* @return CJAX_FRAMEWORK OBJECT
	*/
	public static function getInstance(){
		if(self::$CJAX){
			return self::$CJAX;
		}		
		CoreEvents::errorHandlingConfig();
		$ajax = new self;
		if(!defined('JSON_FORCE_OBJECT')){
			define('JSON_FORCE_OBJECT', 16);
		}
		
		if(!isset($ajax->format) || !$ajax->format){
			$ajax->format = new Format;		
			$config = new Ext;
			if(file_exists($f = CJAX_HOME."/"."config.php")){
				include($f);
				if(isset($config)){
					$config = new Ext($config);
				}
			}
			$ajax->config = $config;
			
			$ajax->initiate($ajax);
			if(!$ajax->isAjaxRequest() && defined('AJAX_CD')){
				@setcookie('AJAX_CD', AJAX_CD, null, '/');
			} 
            else{
				if(isset($_COOKIE['AJAX_CD']) && !defined('AJAX_CD')){
					define('AJAX_CD', $_COOKIE['AJAX_CD']);
				}
				if(!function_exists('cleanInput')){
					function cleanInput($input){
						return $input;
					}
				}
			}
		}
        
		if($ajax->config->ipDebug){
			if(is_array($ajax->config->ipDebug) && in_array(@$_SERVER['REMOTE_ADDR'], $ajax->config->ipDebug)){
				$ajax->config->ipDebug = false;
			} 
            elseif(@$_SERVER['REMOTE_ADDR']!=$ajax->config->ipDebug){
				$ajax->config->ipDebug = false;
			}
		}
		
		if($ajax->config->caching && isset($_SERVER['REQUEST_URI'])){
			$ajax->crc32 = crc32($_SERVER['REQUEST_URI']);
			$cache = $ajax->readCache('cjax-'.$ajax->crc32);
			$ajax->caching = $cache;
		}
		if($ajax->config->debug){
			@ini_set('display_errors', 1);
			@ini_set('log_errors', 1);
		}
		
		if(!$jsDir = $ajax->config->jsPath){
			if(@is_dir('cjax/')){
				$jsDir  = "cjax/assets/js/";
			} 
            elseif(@is_dir('assets/js/')){
				$jsDir  = "assets/js/";
			} 
            elseif(@is_dir('../cjax')){
				$jsDir  = "../cjax/assets/js/";
			} 
            elseif(@is_dir('../../cjax')){
				$jsDir  = "../../cjax/assets/js/";
			} 
            elseif(@is_dir('../../../cjax')){
				$jsDir  = "../../../cjax/assets/js/";
			} 
            else{
				die("Cannot find the correct path to Js directory.");
			}
			
			$error = error_get_last();			
			if($error && preg_match('%.*(open_basedir).*File\(([^\)]+)%', $error['message'], $match)){
				die(sprintf('Restriction <b>open_basedir</b> is turned on. File or directory %s will not be accessible while this setting is on due to security directory range.', $match[2]));
			}
		}
		$ajax->js($jsDir);
		
		return self::$CJAX = $ajax;
	}

    public function initiateRequest(){
        if(!$this->handleModRewrite()) return;
        $this->handleFriendlyURLs();        
    }

    private function handleModRewrite(){
        if(isset($_SERVER['REDIRECT_QUERY_STRING']) && $_SERVER['REDIRECT_QUERY_STRING']){
        	$_SERVER['QUERY_STRING'] = $_SERVER['REDIRECT_QUERY_STRING'];
        } 
        elseif(isset($_SERVER['PATH_INFO']) && $_SERVER['PATH_INFO'] && !$_SERVER['QUERY_STRING']){
	        $_SERVER['QUERY_STRING'] = ltrim($_SERVER['PATH_INFO'],'/');
        }

        $file = 'ajax.php';
        if(isset($_SERVER['SCRIPT_NAME'])){
	        $file = preg_replace("/.+\//",'', ltrim($_SERVER['SCRIPT_NAME'],'/'));
        }
        if(defined('AJAX_FILE') && AJAX_FILE != $file){
	        return false;
        }
        return true;
    }

    private function handleFriendlyURLs(){
        if(isset($_SERVER['QUERY_STRING']) && $query = $_SERVER['QUERY_STRING']){
        	$packet = explode('/' ,rtrim($query,'/'));
        	if(count($packet) == 1) {
                $this->handlePlugin($packet[0]);
	        }
        	if($this->isAjaxRequest() || defined('AJAX_VIEW') ){
                $this->handleController($packet);
	        }
        }

        if(!$this->isAjaxRequest() && count(array_keys(debug_backtrace(false))) == 1 && !defined('AJAX_VIEW')){
	        exit("Security Error. You cannot access this file directly.");
        }
    }

    private function handlePlugin($isPlugin = null){
	    if($plugin = $this->isPlugin($isPlugin) && !defined('AJAX_VIEW')){
		    define('AJAX_VIEW', true);
	    }        
    }

    private function handleController($packet = null){
	    if($packet && count(array_keys($packet)) >= 2 && $packet[0] && $packet[1]){
		    $_REQUEST['controller'] = $packet[0];
		    $_REQUEST['function'] = $packet[1];
		    $_REQUEST['cjax'] = time();
		    if(count(array_keys($packet)) > 2){
			    unset($packet[0]);
				unset($packet[1]);
				if($packet){
				    $params = range('a','z');
					foreach($packet as $k  => $v){
					    $_REQUEST[current($params)] = $v;
						next($params);
				    }
				}
		    }
		} 
        elseif(!$this->input('controller') && count($packet) == 1){
			$url = explode('&',$_SERVER['QUERY_STRING']);
			if(count($url) == 1){
				$_REQUEST['controller'] = $packet[0];
			}
	    }        
    }
}

function ajax(){
	return CJAX::getInstance();
}